### PR TITLE
SDKCF-243 Replace XCTest framework with Kiwi framework for RingBuffer…

### PR DESCRIPTION
**Pull request by vito1188**
… class

### CODE REVIEW

File: Tests/RingBufferTestsKiwi.m [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+SPEC_BEGIN(RingBufferTestsKiwi)
+
+describe(@"RingBuffer", ^{
+
+    describe(@"init", ^{
+
+        it(@"should be non-nil when initialized with non-zero size", ^{
+            _RPTRingBuffer *buffer = [[_RPTRingBuffer alloc] initWithSize:12];
+            [[buffer should] beNonNil];
```
Comment (donnie-jp) [inline Tests/RingBufferTestsKiwi.m]: Please reformat tests like:
setup
_blank line_
action
_blank line_
assertion

File: Tests/RingBufferTestsKiwi.m [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            it(@"should return a non-nil measurement when the passed index is less than the buffer's size", ^{
+                NSUInteger validIndex = size-1;
+                _RPTMeasurement *measurement = [buffer measurementAtIndex:validIndex];
+                [[measurement should] beNonNil];
+            });
+
+            it(@"should return a measurement having the same tracking identifier as expected", ^{
+                NSUInteger validIndex = size-1;
+                _RPTMeasurement *measurement = [buffer measurementAtIndex:validIndex];
+                [[theValue(measurement.trackingIdentifier) should] equal:theValue(validIndex)];
```
Comment (donnie-jp) [inline Tests/RingBufferTestsKiwi.m]: Getting a measurement at validIndex and then comparing trackingIdentifier to validIndex seems confusing.

Also it's not clear why trackingIdentifier should equal validIndex.

File: Tests/RingBufferTestsKiwi.m [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+            it(@"should return a nil measurement when the passed index is equal to the buffer's size", ^{
+                _RPTMeasurement *measurement = [buffer measurementAtIndex:size];
+                [[measurement should] beNil];
+            });
+        });
+
+        context(@"measurementWithTrackingIdentifier", ^{
+
+            it(@"should return a non-nil measurement when the passed an index is less than the buffer's size", ^{
```
Comment (donnie-jp) [inline Tests/RingBufferTestsKiwi.m]: index -> tracking id

File: Tests/RingBufferTestsKiwi.m [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    describe(@"nextMeasurement", ^{
+
+        it(@"should return a non-nil measurement when the buffer is not full", ^{
+            _RPTRingBuffer *buffer = [[_RPTRingBuffer alloc] initWithSize:10];
+            _RPTMeasurement *measurement = nil;
+            for (int i = 0; i < 5; i++) {
+                measurement = [buffer nextMeasurement];
+            }
+            [[measurement should] beNonNil];
+            [[theValue(measurement.trackingIdentifier) should] equal:theValue(5)];
```
Comment (donnie-jp) [inline Tests/RingBufferTestsKiwi.m]: Need to split.
Each test should check 1 assumption.

File: Tests/RingBufferTestsKiwi.m [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        });
+
+        it(@"should return a nil measurement when the buffer is full", ^{
+            _RPTRingBuffer *buffer = [[_RPTRingBuffer alloc] initWithSize:10];
+            _RPTMeasurement *measurement = nil;
+            for (int i = 0; i < 11; i++) {
+                measurement = [buffer nextMeasurement];
+            }
+            [[measurement should] beNil];
+        });
```
Comment (donnie-jp) [inline Tests/RingBufferTestsKiwi.m]: We could also test the scenario where the buffer has been filled, then partially processed e.g. index 0-5 are cleared but 6-9 are filled, calling get nextMeasurement should succeed.

File: Tests/RingBufferTestsKiwi.m [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    });
+
+    describe(@"nextMeasurement", ^{
+
+        it(@"should return a non-nil measurement when the buffer is not full", ^{
+            _RPTRingBuffer *buffer = [[_RPTRingBuffer alloc] initWithSize:10];
+            _RPTMeasurement *measurement = nil;
+            for (int i = 0; i < 5; i++) {
+                measurement = [buffer nextMeasurement];
+            }
```
Comment (donnie-jp) [inline Tests/RingBufferTestsKiwi.m]: I think setup like this can be put into a helper like fillBufferAndReturnNextMeasurement(size, from, to)

**Review by donnie-jp — CHANGES_REQUESTED** [review]
Inline comments

File: Tests/RingBufferTestsKiwi.m [hunk 7]
```diff
@@ -14,113 +15,113 @@ @interface _RPTRingBuffer ()
...
-            [[theValue(measurement.trackingIdentifier) should] equal:theValue(5)];
+        });
+
+        it(@"should return a non-nil measurement having expected tracking identifier when the buffer is not full", ^{
+            _RPTRingBuffer *buffer = [[_RPTRingBuffer alloc] initWithSize:10];
+            mkFillBuffer(buffer, 5, 10);
+
+            _RPTMeasurement *measurement = [buffer nextMeasurement];
+
+            [[theValue(measurement.trackingIdentifier) should] equal:theValue(11)];
```
Comment (donnie-jp) [inline Tests/RingBufferTestsKiwi.m:116]: Can you think of a way to make it clearer why we expect 11 here?

Comment (vito1188) [reply Tests/RingBufferTestsKiwi.m:116]: the buffer is filled by the numbers from 1 to 10, so the next one should be 11.

Comment (donnie-jp) [reply Tests/RingBufferTestsKiwi.m:116]: Yeah I know that but only cos I know how the implementation works. Maybe make the test name clearer like
`should return a non-nil measurement with tracking identifier set to 11 when tracking identifiers up to 10 have already been used`

**Review by donnie-jp — APPROVED** [review]

**Pull request merged**